### PR TITLE
Adds Pagination widget and modifies Table widget (NEW PR)

### PR DESCRIPTION
**Pull request by mlzummo**
@GordonSmith Please Review
Signed-off-by: Mathew Zummo <mzummo@gmail.com>

### CODE REVIEW

File: src/other/Table.js [hunk 1]
```diff
@@ -13,26 +13,15 @@
...
-                [41.244123, -95.961610],
-                [32.688980, -117.192040],
-                [45.786490, -108.526600],
-                [45.796180, -108.535652],
-                [45.774320, -108.494370],
-                [45.777062, -108.549835, "red-dot.png"]
-            ])
-        ;
-        return this;
-    };
```
Comment (GordonSmith) [inline src/other/Table.js]: Don't delete the testData

Comment (mlzummo) [inline src/other/Table.js]: its not ... scrolldown further ... i had added it in my commit long long ago

File: src/other/Table.js [hunk 2]
```diff
@@ -96,5 +125,22 @@
...
+            [ 32.688980, -117.192040 ],
+            [ 45.786490, -108.526600 ],
+            [ 45.796180, -108.535652 ],
+            [ 45.774320, -108.494370 ],
+            [ 45.777062, -108.549835, "red-dot.png" ]
+        ]);
+        
+        return this;
+    };
+
```
Comment (GordonSmith) [inline src/other/Table.js]: If you want to re-factor code (or generally make changes wich have no affect on the logic), you should do it as a separate pull request - as I have no way seeing what changed in testData,,,

File: widgets/src/other/Paginator.css [hunk 3]
```diff
@@ -0,0 +1,93 @@
...
+.pagination-sm > li:first-child > a,
+.pagination-sm > li:first-child > span {
+  border-top-left-radius: 3px;
+  border-bottom-left-radius: 3px;
+}
+.pagination-sm > li:last-child > a,
+.pagination-sm > li:last-child > span {
+  border-top-right-radius: 3px;
+  border-bottom-right-radius: 3px;
+}
```
Comment (GordonSmith) [inline widgets/src/other/Paginator.css]: Style should be inside a ".other_Paginator" namespace.

File: widgets/src/other/Paginator.js [hunk 4]
```diff
@@ -0,0 +1,100 @@
+(function (root, factory) {
+    if (typeof define === "function" && define.amd) {
+        define(["../common/HTMLWidget","css!./Paginator"], factory);
+    } else {
+        root.Entity = factory(root.HTMLWidget);
```
Comment (GordonSmith) [inline widgets/src/other/Paginator.js]: should be root.other_Paginator and root.common_HTMLWidget

File: widgets/src/other/Paginator.js [hunk 5]
```diff
@@ -0,0 +1,100 @@
...
+
+            .on('click', function(d, i) {
+
+                d3.event.preventDefault();
+                
+                if (d=='next') {
+
+                    if ((context.pageNumber()+1) <= context._tNumPages) {
+      
+                        var p = context.pageNumber()+1; context.pageNumber(p);
```
Comment (GordonSmith) [inline widgets/src/other/Paginator.js]: Formatting is confusing - move to new line.

File: widgets/src/other/Paginator.js [hunk 6]
```diff
@@ -0,0 +1,100 @@
...
+                if (d=='next') {
+
+                    if ((context.pageNumber()+1) <= context._tNumPages) {
+      
+                        var p = context.pageNumber()+1; context.pageNumber(p); 
+                        context._onSelect(p,"next"); 
+
+                    }
+
+                } else if (d=='previous') {
```
Comment (GordonSmith) [inline widgets/src/other/Paginator.js]: === preferred

File: widgets/src/other/Paginator.js [hunk 7]
```diff
@@ -0,0 +1,100 @@
...
+                    }
+
+                } else {
+                    context.pageNumber(d); 
+                    context._onSelect(d);
+                }
+            })
+        ;
+            
+        page.classed("active", function(e, j) { return j == context.pageNumber(); })
```
Comment (GordonSmith) [inline widgets/src/other/Paginator.js]: === preferred

**Comment by GordonSmith** [comment]
Lines 53->75 seem to have spurious blank lines...

Also you need to move from /widgets/src -> /src

File: src/other/Table.js [hunk 8]
```diff
@@ -57,7 +46,44 @@
             .remove()
         ;
 
-        var rows = this.tbody.selectAll("tr").data(this._data);
+        if (this.pagination()) {
+
+            if (this._paginator.target() == null) {
+
```
Comment (GordonSmith) [inline src/other/Table.js]: Spurious blank line.

File: src/other/Table.js [hunk 9]
```diff
@@ -57,7 +46,44 @@
             .remove()
         ;
 
-        var rows = this.tbody.selectAll("tr").data(this._data);
+        if (this.pagination()) {
+
```
Comment (GordonSmith) [inline src/other/Table.js]: Spurious blank line.

File: src/other/Table.js [hunk 10]
```diff
@@ -57,7 +46,44 @@
...
+            if (this._paginator.target() == null) {
+
+                var pnode = document.getElementById(this.id()).parentNode.id 
+                this._paginator.target(pnode); 
+
```
Comment (GordonSmith) [inline src/other/Table.js]: Spurious blank line.

File: src/other/Table.js [hunk 11]
```diff
@@ -57,7 +46,44 @@
...
+            this._tNumPages = Math.ceil(this._numItems() / this.itemsPerPage()) || 1; // TODO make this into a function in paginator?
+            if (this.pageNumber() > this._tNumPages ) { this.pageNumber(1); } // resets if current pagenum selected out of range
+
+            this._paginator._onSelect = function(p, d) {
+                console.log('page: '+p);
+                // context.pageNumber(p); // technically not needed due to tight coupling of widget
+                context.render();
+                return;
+            };
+
```
Comment (GordonSmith) [inline src/other/Table.js]: Spurious blank line.

File: src/other/Paginator.js [hunk 12]
```diff
@@ -0,0 +1,91 @@
+(function (root, factory) {
+    if (typeof define === "function" && define.amd) {
+        define(["../common/HTMLWidget","css!./Paginator"], factory);
+    } else {
+        root.other_Paginator = factory(root.common_HTMLWidget);
```
Comment (GordonSmith) [inline src/other/Paginator.js]: Missing root paginator

**Comment by GordonSmith** [comment]
Looks fine.

**Comment by buunguyen** [comment]
@mlzummo @GordonSmith is there an example of enabling pagination for table widget? In a previous release (I think candidate-0.0.9), I was able to invoke `table.pagination(true)` to enable pagination. But today when I pulled 1.0.0, from Bower, my code no longer works. The table is shown, but the pages aren't shown.

**Comment by mlzummo** [comment]
http://rawgit.com/hpcc-systems/Visualization/v1.0.0/demos/dermatology.html?src/other/Table
its working there... if you msg me on Google Talk i can help you further you might need to enter itemsPerPage and PageNumber

**Comment by buunguyen** [comment]
@mlzummo thanks a bunch, the demo works fine, turns out I messed up when refactoring code after pulling in candidate-1.0.0.
